parlai/agents/coopgame_agent/coopgame_agent.py

```diff
@@ -0,0 +1,286 @@
+                                'rate (as specified by -lr).')
+        group.add_argument('--learning-rate', default=1e-2, type=float,
+                           help='Initial learning rate')
+        group.add_argument('--no-cuda', action='store_true', default=False,
+                           help='disable GPUs even if available')
+        group.add_argument('--gpuid', type=int, default=-1,
+                           help='which GPU device to use (defaults to cpu)')
+
+    def __init__(self, opt, shared=None):
+        super().__init__(opt, shared)
+        self.id = 'CooperativeGameAgent'
+        self.actions = []
+
+        # basic modules for listening, state update and speaking
+        # questioner will have `PredictNet`, answerer will have `ImgNet`
+        self.listen_net = ListenNet(opt['in_vocab_size'], opt['embed_size'])
+        self.state_net = StateNet(opt['embed_size'], opt['state_size'])
+        self.speak_net = SpeakNet(opt['state_size'], opt['out_vocab_size'])
+
+        # initialize short (h) and long (c) term states
+        self.reset()
+
+        # transfer agent to GPU if applicable
+        self.use_cuda = not opt.get('no_cuda') and torch.cuda.is_available()
+        if self.use_cuda:
+            print('[ Using CUDA for %s ]' % self.id)
+            torch.cuda.set_device(opt['gpuid'])
+            for module in self.modules:
+                module = module.cuda()
+
+        # setup dictionary agent
+        self.dict_agent = dictionary_class()()
+
+        # setup optimizer according to command-line arguments
+        self.optimizer = self.setup_optimizer()
+
+    @property
+    def modules(self):
+        """Property to return a list of pytorch modules. Override this method
+        while subclassing, if extra modules are added (for example, image
+        feature extractor in answerer).
+        """
+        return [self.listen_net, self.state_net, self.speak_net]
+
+    def setup_optimizer(self):
+        """Return a ``torch.nn.optim.optimizer`` according to command-line
+        argument ``--optimizer``. Override this method to setup optimizer with
+        non-default parameters or use custom optimizer not available as choice.
+        """
+        optim_class = CooperativeGameAgent.OPTIM_OPTS[opt['optimizer']]
+        kwargs = {'lr': self.opt['learning_rate']}
+        if opt['optimizer'] == 'sgd':
+            kwargs['momentum'] = 0.95
+            kwargs['nesterov'] = True
+        return optim_class([module.parameters() for module in self.modules],
+                            **kwargs)
+
+    def tokenize(self, text):
+        """Convert text observaton (string) to a ``torch.autograd.Variable`` of
+        tokens using ``DictionaryAgent``.
+        """
+        text_tokens = self.dict.txt2vec(text)
+        # TODO(kd): handle requires_grad=False for valid/test
+        token_vars = Variable(torch.Tensor(text_tokens), requires_grad=True)
```

Comment:
do you need `requires_grad=True` here? maybe I'm missing something, but isn't that only true if you're passing gradients back into the dictionary?

Reply:
I think you're right. I am correcting this in upcoming commit. Also, I added an in-place operation on Variable - removing it.